File: web/src/lib/redaction.ts
/**
 * Client-side redaction of a lab report's text layer before anything leaves the phone
 * (CLAUDE.md rule 2; docs/contracts.md section 7). A port of Lane D's reference applier
 * `api/app/redaction.py`; the rules themselves are D's `api/redaction_rules.json`, copied to
 * `public/redaction_rules.json` by scripts/sync-assets.js and fetched at runtime.
 */

export interface RawRule {
  id: string;
  category: 'name' | 'dob' | 'mrn' | 'address' | 'physician' | 'contact';
  action: 'drop_line' | 'mask';
  pattern: string;
  note?: string;
}

export interface RawRules {
  version: number;
  mask_token?: string;
  keep_if: { pattern: string; note?: string };
  rules: RawRule[];
}

export interface Rule {
  id: string;
  category: RawRule['category'];
  action: RawRule['action'];
  regex: RegExp;
}

export interface RuleSet {
  version: number;
  keepIf: RegExp;
  maskToken: string;
  rules: Rule[];
}

export interface MaskHit {
  ruleId: string;
  matched: string;
  /** [start, end) in the line as it was when the rule ran (masks apply right to left, so earlier offsets stay valid). */
  start: number;
  end: number;
}

export interface LineRedaction {
  /** The line after masking, or null when a drop_line rule removed it. */
  text: string | null;
  droppedBy: string | null;
  masks: MaskHit[];
}

export function loadRules(raw: RawRules): RuleSet {
  return {
    version: raw.version,
    keepIf: new RegExp(raw.keep_if.pattern, 'i'),
    maskToken: raw.mask_token ?? '[REDACTED]',
    rules: raw.rules.map((r) => ({ id: r.id, category: r.category, action: r.action, regex: new RegExp(r.pattern, 'i') })),
  };
}

/** Rules run in order; the first drop_line wins (never on a result row); mask rules replace only the match. */
export function redactLine(line: string, rs: RuleSet): LineRedaction {
  // A line that already carries the mask token was redacted upstream (the consented fixture, a
  // report redacted once before): its label ("Patient:", "DOB:") no longer sits next to a value,
  // so dropping it would only change the bytes and defeat the API's cached response for that file.
  const alreadyRedacted = line.includes(rs.maskToken);
  const isResultRow = rs.keepIf.test(line);
  const masks: MaskHit[] = [];
  let out = line;
  for (const rule of rs.rules) {
    if (rule.action === 'drop_line') {
      if (!isResultRow && !alreadyRedacted && rule.regex.test(out)) return { text: null, droppedBy: rule.id, masks };
    } else {
      const global = new RegExp(rule.regex.source, 'gi');
      const found = [...out.matchAll(global)].filter((m) => m[0].length > 0);
      for (const m of found.reverse()) {
        const start = m.index ?? 0;
        masks.push({ ruleId: rule.id, matched: m[0], start, end: start + m[0].length });
        out = out.slice(0, start) + rs.maskToken + out.slice(start + m[0].length);
      }
    }
  }
  return { text: out, droppedBy: null, masks };
}

export interface TextRedaction {
  text: string;
  dropped: { line: number; ruleId: string; text: string }[];
  masked: { line: number; ruleId: string; matched: string }[];
}

export function redactText(text: string, rs: RuleSet): TextRedaction {
  const kept: string[] = [];
  const result: TextRedaction = { text: '', dropped: [], masked: [] };
  text.split('\n').forEach((line, i) => {
    const r = redactLine(line, rs);
    if (r.text === null) {
      result.dropped.push({ line: i, ruleId: r.droppedBy!, text: line });
      return;
    }
    for (const m of r.masks) result.masked.push({ line: i, ruleId: m.ruleId, matched: m.matched });
    kept.push(r.text);
  });
  result.text = kept.join('\n');
  return result;
}

export function redactionHits(r: TextRedaction): number {
  return r.dropped.length + r.masked.length;
}

File: api/app/redaction.py
"""Reference implementation of api/redaction_rules.json.

Lane C applies the same rules in the browser (pdf.js text items); this module exists so the
rules are tested against fixtures here and so the server can assert an upload was redacted.
The regexes are written in the ECMAScript subset that Python's `re` also accepts.
"""
import json
import re
from dataclasses import dataclass, field
from functools import lru_cache
from pathlib import Path

from .config import API_DIR

RULES_PATH = API_DIR / "redaction_rules.json"


@dataclass(frozen=True)
class Rule:
    id: str
    category: str
    action: str  # "drop_line" | "mask"
    regex: re.Pattern


@dataclass(frozen=True)
class RuleSet:
    version: int
    keep_if: re.Pattern
    mask_token: str
    rules: tuple[Rule, ...]


@dataclass
class Redaction:
    text: str
    dropped: list[tuple[int, str, str]] = field(default_factory=list)  # (line_no, rule_id, line)
    masked: list[tuple[int, str, str]] = field(default_factory=list)  # (line_no, rule_id, match)

    @property
    def hits(self) -> int:
        return len(self.dropped) + len(self.masked)


@lru_cache
def load_rules(path: Path = RULES_PATH) -> RuleSet:
    raw = json.loads(Path(path).read_text(encoding="utf-8"))
    rules = tuple(
        Rule(r["id"], r["category"], r["action"], re.compile(r["pattern"], re.IGNORECASE)) for r in raw["rules"]
    )
    return RuleSet(
        version=raw["version"],
        keep_if=re.compile(raw["keep_if"]["pattern"], re.IGNORECASE),
        mask_token=raw.get("mask_token", "[REDACTED]"),
        rules=rules,
    )


def redact_line(line: str, rs: RuleSet) -> tuple[str | None, list[str], list[tuple[str, str]]]:
    """Returns (new_line or None if dropped, drop rule ids, [(mask rule id, matched text)])."""
    is_result_row = bool(rs.keep_if.search(line))
    masks: list[tuple[str, str]] = []
    out = line
    for rule in rs.rules:
        if rule.action == "drop_line":
            if not is_result_row and rule.regex.search(out):
                return None, [rule.id], masks
        elif rule.action == "mask":
            for m in list(rule.regex.finditer(out))[::-1]:  # right to left so offsets stay valid
                masks.append((rule.id, m.group(0)))
                out = out[: m.start()] + rs.mask_token + out[m.end() :]
    return out, [], masks


def redact_text(text: str, rs: RuleSet | None = None) -> Redaction:
    rs = rs or load_rules()
    result = Redaction(text="")
    kept: list[str] = []
    for i, line in enumerate(text.split("\n")):
        new, drops, masks = redact_line(line, rs)
        if new is None:
            result.dropped.append((i, drops[0], line))
            continue
        for rule_id, matched in masks:
            result.masked.append((i, rule_id, matched))
        kept.append(new)
    result.text = "\n".join(kept)
    return result
